utils/common/src/main/kotlin/EnvironmentVariableFilter.kt

```diff
@@ -0,0 +1,112 @@
+ * variables, such as database or service credentials.
+ *
+ * To reduce this risk, this object filters the environment variables passed to child processes based on the
+ * following criteria:
+ * - Substrings for variable names can be defined to determine variables with sensitive information. The object provides
+ *   some default strings to match variable names like "PASS", "USER", "TOKEN", etc.
+ * - There is an allow list to include variables even if they contain one of these substrings.
+ *
+ * So in order to determine whether a specific variable "E" can be passed to a child process, this filter applies the
+ * following steps:
+ * - If E is contained in the allow list, it is included.
+ * - Otherwise, E is included if and only if its name does not contain one of the exclusion substrings (ignoring case).
+ */
+object EnvironmentVariableFilter {
+    /**
+     * A set with substrings contained in variable names that are denied by default. All variables containing one of
+     * these strings (ignoring case) are not propagated to child processes.
+     */
+    val DEFAULT_DENY_SUBSTRINGS = setOf(
+        "key",
+        "pass",
+        "pwd",
+        "token",
+        "user"
+    )
+
+    /** A set of known variable names that are allowed despite being matched by deny substrings. */
+    val DEFAULT_ALLOW_NAMES = setOf(
+        "CARGO_HTTP_USER_AGENT",
+        "COMPOSER_ALLOW_SUPERUSER",
+        "CONAN_LOGIN_ENCRYPTION_KEY",
+        "CONAN_LOGIN_USERNAME",
+        "CONAN_PASSWORD",
+        "CONAN_USERNAME",
+        "CONAN_USER_HOME",
+        "CONAN_USER_HOME_SHORT",
+        "DOTNET_CLI_CONTEXT_ANSI_PASS_THRU",
+        "GIT_ASKPASS",
+        "GIT_HTTP_USER_AGENT",
+        "GRADLE_USER_HOME",
+        "HACKAGE_USERNAME",
+        "HACKAGE_PASSWORD",
+        "HACKAGE_KEY",
+        "PWD",
+        "USER",
+        "USERPROFILE"
+    )
+
+    /** Stores the current deny substrings used by this filter. */
+    private var denySubstrings = DEFAULT_DENY_SUBSTRINGS
+
+    /** Stores the current allow names used by this filter. */
+    private var allowNames = DEFAULT_ALLOW_NAMES
+
+    /**
+     * Reset this filter to use the given [denySubstrings] and [allowNames].
+     */
+    fun reset(
+        denySubstrings: Collection<String> = DEFAULT_DENY_SUBSTRINGS,
+        allowNames: Collection<String> = DEFAULT_ALLOW_NAMES
+    ) {
+        this.denySubstrings = denySubstrings.toSet()
+        this.allowNames = allowNames.toSet()
+    }
+
+    /**
+     * Test whether the variable with the given [name] can be passed to a child process according to the criteria
+     * described in the header comment.
+     */
+    fun isAllowed(name: String): Boolean {
+        return name in allowNames || denySubstrings.none { name.contains(it, ignoreCase = true) }
+    }
+
+    /**
+     * Remove all keys from [environment] that do not pass this filter.
+     */
+    fun filter(environment: MutableMap<String, String>): MutableMap<String, String> {
```

Comment:
Having a function that modifies it parameters AND returns the modified version is a bit of codesmell.

I would rather have either an extension function `MutableMap<String, String>.filter` or no return value.

Reply:
I like the idea of an extension function.

Reply:
The parameter is returned to have a single expression in `ProcessCapture` where the filter is applied. I am not sure whether I understand the proposal for an extension function. How would the signature of such an extension function look like and how would it help in this context?

Reply:
As written `MutableMap<String, String>.filter()` and applied like that:
```
with(EnvironmentVariableFilter) {
    environment().filter().putAll(environment)
}
```
but it is more verbose.

Reply:
Or remove the return value, you don't use it in `ProcessCapture`.

Reply:
You're right: since `environment()` is a `MutableMap` and there is no 'cleaner' way of changing it, I am find to leave the code as it is.